src/Math/src/CARE.cpp

```diff
@@ -0,0 +1,191 @@
+/**
+ * @file CARE.cpp
+ * @authors Giulio Romualdi
+ * @copyright 2020 Istituto Italiano di Tecnologia (IIT). This software may be modified and
+ * distributed under the terms of the GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#include <iostream>
+#include <sstream>
+
+#include <BipedalLocomotion/Math/CARE.h>
+
+using namespace BipedalLocomotion::Math;
+
+struct CARE::Impl
+{
+    double tolerance{1e-9};
+    bool isVerbose{false};
+    int maxIterations{100};
+
+    Eigen::MatrixXd A;
+    Eigen::MatrixXd B;
+    Eigen::MatrixXd Q;
+    Eigen::MatrixXd R;
+
+    Eigen::MatrixXd lhs;
+    Eigen::MatrixXd rhs;
+    Eigen::MatrixXd Z;
+    Eigen::MatrixXd ZOld;
+
+    Eigen::MatrixXd identity;
+    Eigen::MatrixXd solution;
+
+    void resizeInteralMatrices()
+    {
+        const std::size_t n = B.rows();
+
+        // resize the Hamiltonian matrix
+        Z.resize(2 * n, 2 * n);
+        ZOld.resize(2 * n, 2 * n);
+
+        identity = Eigen::MatrixXd::Identity(n, n);
```

Comment:
Here I would use a ``resize`` and a ``setIdentity``. In this way, if the size is already the correct one, it avoids memory allocations.

Reply:
It should be done in b320996